smithy-typescript-codegen/src/main/java/software/amazon/smithy/typescript/codegen/TypeScriptJmesPathVisitor.java

```diff
@@ -0,0 +1,315 @@
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package software.amazon.smithy.typescript.codegen;
+
+import java.util.ArrayList;
+
+import software.amazon.smithy.codegen.core.CodegenException;
+import software.amazon.smithy.jmespath.ExpressionVisitor;
+import software.amazon.smithy.jmespath.JmespathExpression;
+import software.amazon.smithy.jmespath.ast.AndExpression;
+import software.amazon.smithy.jmespath.ast.ComparatorExpression;
+import software.amazon.smithy.jmespath.ast.CurrentExpression;
+import software.amazon.smithy.jmespath.ast.ExpressionTypeExpression;
+import software.amazon.smithy.jmespath.ast.FieldExpression;
+import software.amazon.smithy.jmespath.ast.FilterProjectionExpression;
+import software.amazon.smithy.jmespath.ast.FlattenExpression;
+import software.amazon.smithy.jmespath.ast.FunctionExpression;
+import software.amazon.smithy.jmespath.ast.IndexExpression;
+import software.amazon.smithy.jmespath.ast.LiteralExpression;
+import software.amazon.smithy.jmespath.ast.MultiSelectHashExpression;
+import software.amazon.smithy.jmespath.ast.MultiSelectListExpression;
+import software.amazon.smithy.jmespath.ast.NotExpression;
+import software.amazon.smithy.jmespath.ast.ObjectProjectionExpression;
+import software.amazon.smithy.jmespath.ast.OrExpression;
+import software.amazon.smithy.jmespath.ast.ProjectionExpression;
+import software.amazon.smithy.jmespath.ast.SliceExpression;
+import software.amazon.smithy.jmespath.ast.Subexpression;
+
+class TypeScriptJmesPathVisitor implements ExpressionVisitor<Void> {
+
+    TypeScriptWriter writer;
+    String executionContext;
+
+    String accessor;
+    JmespathExpression jmesExpression;
+    int scopeCount;
+
+    TypeScriptJmesPathVisitor(TypeScriptWriter writer, String accessor, JmespathExpression expression) {
+        this.writer = writer;
+        this.accessor = accessor;
+        this.executionContext = accessor;
+        this.jmesExpression = expression;
+        scopeCount = 0;
+    }
+
+    private String makeNewScope(String prefix) {
+        scopeCount += 1;
+        return prefix + scopeCount;
+    }
+
+    private static boolean isNumeric(String strNum) {
+        if (strNum == null) {
+            return false;
+        }
+        try {
+            Double.parseDouble(strNum);
+        } catch (NumberFormatException nfe) {
+            return false;
+        }
+        return true;
+    }
+
+    public void run() {
+        writer.openBlock("let returnComparer = () => {", "}", () -> {
+            executionContext = this.accessor;
+            jmesExpression.accept(this);
+            writer.write("return $L;", this.executionContext);
+        });
+        executionContext = "returnComparer()";
```

Comment:
What is a comparer? In JMESPath we call these comparator if that helps.

Reply:
just a name, i'll change to returnComparator

Reply:
this is what the acceptor condition checks against